core/kernel/tee_ta_manager.c

```diff
@@ -823,6 +843,132 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
 	return res;
 }
 
+#if defined(CFG_TA_STATS)
+static TEE_Result tee_ta_dump_memstats(struct tee_ta_session *s,
+				       uint32_t cancel_req_to,
+				       struct tee_ta_param *param)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_ctx *ctx = NULL;
+	struct ts_ctx *ts_ctx = NULL;
+
+	if (!check_params(s, param))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ts_ctx = s->ts_sess.ctx;
+	if (ts_ctx)
+		ctx = ts_to_ta_ctx(ts_ctx);
+
+	if (ctx->panicked) {
+		DMSG("Panicked !");
+		return TEE_ERROR_TARGET_DEAD;
+	}
+
+	if (tee_ta_try_set_busy_nolock(ctx)) {
+		s->param = param;
+		set_invoke_timeout(s, cancel_req_to);
+		res = ts_ctx->ops->dump_mem_stats(&s->ts_sess);
+		tee_ta_clear_busy_nolock(ctx);
+	} else {
+		/* Deadlock avoided */
+		res = TEE_ERROR_BUSY;
+	}
+	s->param = NULL;
+
+	return res;
+}
+
+TEE_Result tee_ta_instance_stats(void *buff,
+				 uint32_t *buff_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_param param;
+	struct tee_ta_dump_stats *dump_stats = NULL;
+	struct tee_ta_session *sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_ctx *ctx = NULL;
+	uint32_t count = 0;
+	uint32_t size = 0;
+
+	if (!buff_size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&tee_ta_mutex);
+	if (!buff || *buff_size == 0) {
+		/*
+		 * Buff is null or pass size with 0 means caller want to get
+		 * the size of buff.
+		 * Go through all available TA and calc out the buffer size.
+		 */
+		TAILQ_FOREACH(ctx, &tee_ctxes, link) {
+			count++;
+		}
+		*buff_size = sizeof(struct tee_ta_dump_stats) * count;
+		goto out;
+	}
+
+	dump_stats = (struct tee_ta_dump_stats *)buff;
+	size = *buff_size;
+	nsec_sessions_list_head(&open_sessions);
+	/*
+	 * Scan all sessions opened from secure side by searching through
+	 * all available TA instances and for each context, scan all opened
+	 * sessions.
+	 */
+	/* Context of a loaded TA */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link) {
+		if (size < sizeof(struct tee_ta_dump_stats)) {
+			res = TEE_ERROR_SHORT_BUFFER;
```

Comment:
should update `*buff_size` to output the required size.
Note that if new context are created by the time caller recalls with appropriate size, the last call will also report a SHORT_BUFFER error. Caller should be aware of that.

Reply:
good point.